deepecho/models/basic_gan.py

```diff
@@ -0,0 +1,575 @@
+                to the actual time series data such that `data[i][j]` contains
+                the value at the jth time step of the ith channel of the
+                multivariate time series.
+            context_types:
+                List of strings indicating the type of each value in context.
+                he value at `context[i]` must match the type specified by
+                `context_types[i]`. Valid types include the following: `categorical`,
+                `continuous`, `ordinal`, `count`, and `datetime`.
+            data_types:
+                List of strings indicating the type of each channel in data.
+                Each value in the list at data[i] must match the type specified by
+                `data_types[i]`. The valid types are the same as for `context_types`.
+        """
+        self._analyze_data(sequences, context_types, data_types)
+
+        data = self._build_tensor(self._data_to_tensor, sequences, 'data', dim=1)
+        context = self._build_tensor(self._context_to_tensor, sequences, 'context', dim=0)
+        data_context = _expand_context(data, context)
+
+        discriminator, generator_opt, discriminator_opt = self._build_fit_artifacts()
+
+        iterator = range(self._epochs)
+        if self._verbose:
+            iterator = tqdm(iterator)
+
+        for epoch in iterator:
+            discriminator_score = self._discriminator_step(
+                discriminator=discriminator,
+                discriminator_opt=discriminator_opt,
+                data_context=data_context,
+                context=context,
+            )
+            generator_score = self._generator_step(
+                discriminator=discriminator,
+                generator_opt=generator_opt,
+                context=context,
+            )
+
+            if self._verbose:
+                iterator.set_description(
+                    'Epoch {} | D Loss {} | G Loss {}'.format(
+                        epoch + 1, discriminator_score.item(), generator_score.item()
+                    )
+                )
+
+    def sample_sequence(self, context, sequence_length=None):
+        """Sample a single sequence conditioned on context.
+
+        Args:
+            context (list):
+                The list of values to condition on. It must match
+                the types specified in context_types when fit was called.
+
+        Returns:
+            list[list]:
+                A list of lists (data) corresponding to the types specified
+                in data_types when fit was called.
+        """
+        context = self._context_to_tensor(context).unsqueeze(0).to(self._device)
+
+        with torch.no_grad():
+            generated = self._generate(context, sequence_length)
+            if sequence_length is None:
+                end_flag = generated[:, 0, -1]
+                if (end_flag == 1.0).any():
+                    cut_index = end_flag.cpu().numpy().argmax()
```

Comment:
I think the logic here should be to find the index of the first value that is greater than 0.5 and use that as the `cut_index`, rather than the index of the maximum. For example, if we have:

> [0.0, 0.0, 0.6, 0.7, 0.8]

It should be cut to

> [0.0, 0.0, 0.6]

Reply:
Bear in mind that the values have been rounded before this point is reached.
So the example vector `[0.0, 0.0, 0.6, 0.7, 0.8]` is gets here actually as `[0.0, 0.0, 1.0, 1.0, 1.0]`.

So the `argmax` really returns the "First index which is 1.0", which is the one we are after.
Btw: notice that I move the array to CPU and convert to numpy. This is because the `torch.argmax` output is not guaranteed to return the "first occurrence" when the tensor is in a GPU, but `numpy.argmax` does.
This actually introduces some overhead that makes this step slower, so if we can thing of a way to get the first occurrence that is purely GPU based, we might get a noticeable performance improvement. I'd leave this for a future PR.